Sorted vectors have intersection

amcastro-tri opened the pull request:
This PR pulls out the implementation for SortedVectorsHaveIntersection from the old reverted PR #2299.

This algorithm (placed in drake/common) will be used in the implementation of collision groups (#1793) but can however be reviewed standalone. 

PR #2299 had been reviewed and merged already (however reverted due to Gurobi issues) so I'd expect this snippet of code to be easily re-reviewed this time.

This PR is a step towards resolving #2709, #2710.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2863)
<!-- Reviewable:end -->

amcastro-tri commented:
+@sherm1 for feature review since you reviewed #2299 last time.
+@sammy-tri  for platform review since it is Wednesday. However, feel free to re-assign to @ggould-tri  since he was the platform reviewer last time.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KN8ddrwLholSMG7nvOD:1858652399)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/test/CMakeLists.txt, line 9 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8oyzGjdD8h16U1tYa:-KN8oyzHVK5VjRkoAuxU:-1056488936) ([raw file](https://github.com/robotlocomotion/drake/blob/7d72cec73427303415481b7d5e4c7c0946f51164/drake/common/test/CMakeLists.txt#L9)):*
> ```CMake
> add_executable(have_intersection_test have_intersection_test.cc)
> target_link_libraries(have_intersection_test
>         drakeCommon ${GTEST_BOTH_LIBRARIES})
> ```

Does the test actually use anything from drakeCommon?  It seems to be testing a header only implementation (which I guess is related to drakeCommon...)

---

*[drake/common/test/have_intersection_test.cc, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8pT_2YPuIIbp5VMGm:-KN8pT_3ZOadV4nVwMnH:73789549) ([raw file](https://github.com/robotlocomotion/drake/blob/7d72cec73427303415481b7d5e4c7c0946f51164/drake/common/test/have_intersection_test.cc#L4)):*
> ```cpp
> 
> #include <vector>
> #include <set>
> ```

Include appears to be unused

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/test/CMakeLists.txt, line 9 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8oyzGjdD8h16U1tYa:-KN8tpnIWwr8Gizb4zfy:771310362) ([raw file](https://github.com/robotlocomotion/drake/blob/7d72cec73427303415481b7d5e4c7c0946f51164/drake/common/test/CMakeLists.txt#L9)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Does the test actually use anything from drakeCommon?  It seems to be testing a header only implementation (which I guess is related to drakeCommon...)
</details>
Good point. Removed drakeCommon.

---

*[drake/common/test/have_intersection_test.cc, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8pT_2YPuIIbp5VMGm:-KN8tsl4B3hnVB9KFnuj:-1424288426) ([raw file](https://github.com/robotlocomotion/drake/blob/7d72cec73427303415481b7d5e4c7c0946f51164/drake/common/test/have_intersection_test.cc#L4)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Include appears to be unused
</details>
Removed, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/test/have_intersection_test.cc, line 7 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8xowtwdXgNetAiutq:-KN8zBaVjBWcu9C8EaGo:-1363948229) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/test/have_intersection_test.cc#L7)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Using appears to be unused (I expected it to error when the include was removed, but I guess it's leaking in from somewhere else...)
</details>
Very nice catch. Interesting that it compiles.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KN9-7Jg5_5IM3lNJYxK:1417088721)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Just a few minor comments ...

---

Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[drake/common/sorted_vectors_have_intersection.h, line 18 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8vY7qvHxRE_HFrKNc:-KN8vY7qvHxRE_HFrKNd:-1804376164) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/sorted_vectors_have_intersection.h#L18)):*
> ```objc
> 
> This algorithm only works on `std::vector`'s to take advantage of their fast
>  and random access.
> ```

extra space at beginning of line

---

*[drake/common/sorted_vectors_have_intersection.h, line 37 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8yzJRX_u2_XHgdS1C:-KN8yzJRX_u2_XHgdS1D:1666139410) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/sorted_vectors_have_intersection.h#L37)):*
> ```objc
>   typename std::vector<T>::const_iterator ai = a.begin();
>   typename std::vector<T>::const_iterator bi = b.begin();
>   // Quick checks first:
> ```

In Debug builds I think this should check the precondition that the lists are sorted. This would work:
```c++
#include <algorithm>
// ...
DRAKE_ASSERT(std::is_sorted(a.cbegin(), a.cend()) &&
             std::is_sorted(b.cbegin(), b.cend()));
```
That mean O(n) performance in Debug builds but could also catch some very obscure errors. I don't think the performance hit would be noticeable in Debug for the kinds of short lists we're expecting.

---

*[drake/common/test/CMakeLists.txt, line 7 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8wMtEWUQhLEmskd-D:-KN8wMtEWUQhLEmskd-E:-1465363567) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/test/CMakeLists.txt#L7)):*
> ```CMake
> add_test(NAME nice_type_name_test COMMAND nice_type_name_test)
> 
> add_executable(have_intersection_test have_intersection_test.cc)
> ```

For consistency, this should be called `sorted_vectors_have_intersection_test`.

---

*[drake/common/test/have_intersection_test.cc, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9-jxvelEL9vDNiVtW:-KN9-jxvelEL9vDNiVtX:1936149198) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/test/have_intersection_test.cc#L1)):*
> ```cpp
> #include "drake/common/sorted_vectors_have_intersection.h"
> ```

This file should be renamed to match the header name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KN9-xEw9ZVIfvI28X5Q:1074879139)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/sorted_vectors_have_intersection.h, line 18 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8vY7qvHxRE_HFrKNc:-KN95f3w5kJyEL6R_ns3:1675637715) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/sorted_vectors_have_intersection.h#L18)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> extra space at beginning of line
</details>
removed.

---

*[drake/common/sorted_vectors_have_intersection.h, line 37 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8yzJRX_u2_XHgdS1C:-KN9AA0eFXVcjBxYzaW8:239585919) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/sorted_vectors_have_intersection.h#L37)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> In Debug builds I think this should check the precondition that the lists are sorted. This would work:
> ```c++
> #include <algorithm>
> // ...
> DRAKE_ASSERT(std::is_sorted(a.cbegin(), a.cend()) &&
>              std::is_sorted(b.cbegin(), b.cend()));
> ```
> That mean O(n) performance in Debug builds but could also catch some very obscure errors. I don't think the performance hit would be noticeable in Debug for the kinds of short lists we're expecting.
</details>
Check added.

---

*[drake/common/test/CMakeLists.txt, line 7 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN8wMtEWUQhLEmskd-D:-KN96h7Rr7XxnuIRL-8T:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/test/CMakeLists.txt#L7)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> For consistency, this should be called `sorted_vectors_have_intersection_test`.
</details>
Done.

---

*[drake/common/test/have_intersection_test.cc, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9-jxvelEL9vDNiVtW:-KN96hnj548yFLNVbECD:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/26568bdfeb96769d9c1eec5d4ef357fdaa7a834f/drake/common/test/have_intersection_test.cc#L1)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This file should be renamed to match the header name.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 5 files at r1, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/common/sorted_vectors_have_intersection.h, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9ohiksewiwuW06m-9:-KN9ohiksewiwuW06m-A:-1273878808) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L4)):*
> ```objc
> 
> #include <vector>
> #include <algorithm>
> ```

These should be alphabetically sorted.

---

*[drake/common/sorted_vectors_have_intersection.h, line 34 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9qJYEDH5qpO90teoY:-KN9qJYFaDkESpYZytWM:1457958201) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L34)):*
> ```objc
> @endverbatim
> 
> In this case the algorithm needs to scan both vectors from start to end. **/
> ```

Missing API documentation on the requirements of `@tparam T`.

---

*[drake/common/sorted_vectors_have_intersection.h, line 62 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9oJDA6SNQaUhYFsdk:-KN9oJDA6SNQaUhYFsdl:-516175535) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L62)):*
> ```objc
>       if (++bi == b.end()) break;
>     } else {
>       // Found matching element.
> ```

I was going to suggest `DRAKE_ASSERT(*ai == *bi);` for readability and robustness.  (That way, for example, `NaN` won't intersect all-everything.)  But that adds a requirement on `T` to have `operator==` instead of just `operator<`, so I guess it's not worth it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/common/sorted_vectors_have_intersection.h, line 34 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9qJYEDH5qpO90teoY:-KNAQbXMhKe7IM45LWMf:-1034651697) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L34)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing API documentation on the requirements of `@tparam T`.
</details>
Per comment below this documentation should at least warn about problems with floating point types, if not ban them altogether.

---

*[drake/common/sorted_vectors_have_intersection.h, line 62 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9oJDA6SNQaUhYFsdk:-KNAOrDV6Mav_snNcIrU:1052570893) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L62)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I was going to suggest `DRAKE_ASSERT(*ai == *bi);` for readability and robustness.  (That way, for example, `NaN` won't intersect all-everything.)  But that adds a requirement on `T` to have `operator==` instead of just `operator<`, so I guess it's not worth it.
</details>
Good observation, Jeremy. This algorithm would be a poor choice to use for floating point types T since it (a) requires perfect bit-pattern matches, and (b) fails on NaN as Jeremy noted. The `@tparam T` documentation at least should point that out, but also consider a `static_assert` to prevent use with floating types.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
This had a real failure in CI in the Debug build:
```
17:23:18 common/test/CMakeFiles/sorted_vectors_have_intersection_test.dir/sorted_vectors_have_intersection_test.cc.o:
In function `bool drake::SortedVectorsHaveIntersection<int>(std::vector<int, std::allocator<int> > const&, std::vector<int, std::allocator<int> > const&)':
17:23:18 /home/ubuntu/workspace/experimental/2e937e5a/drake/pod-build/../../drake/common/sorted_vectors_have_intersection.h:44: 
undefined reference to `drake::detail::Abort(char const*, char const*, char const*, int)'
```
The test has to link with drakeCommon in order to use DRAKE_ASSERT (which calls `drake::detail::Abort`).

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KNASmoQZlA3PycLd8ZS:392200241)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/common/sorted_vectors_have_intersection.h, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9ohiksewiwuW06m-9:-KND3xMT7ewBsqbmx9QQ:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> These should be alphabetically sorted.
</details>
Done.

---

*[drake/common/sorted_vectors_have_intersection.h, line 62 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9oJDA6SNQaUhYFsdk:-KND8Yc654O2CXma7YZL:-1160208653) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L62)):*
Given what this method is indented to do I think it is not a bad idea to require `T` having an `operator==`. How can I require this?
If you look at the (glibc) implementation of `std::set_intersection` they only require `T` to have a `operator<` with some wacky template meta programming.  In the [documentation](http://www.cplusplus.com/reference/algorithm/set_intersection/) it says that:
> Two elements, a and b are considered equivalent if (!(a<b) && !(b<a)) or if (!comp(a,b) && !comp(b,a)).

Maybe documenting this requirement plus the `DRAKE_ASSERT` suggested by @jwnimmer-tri would be enough given we do not have (I think) the machinery to require a type having a given operator?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/sorted_vectors_have_intersection.h, line 62 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9oJDA6SNQaUhYFsdk:-KNDC9WOOLtQUaGSSFg3:-1530202319) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L62)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Given what this method is indented to do I think it is not a bad idea to require `T` having an `operator==`. How can I require this?
> If you look at the (glibc) implementation of `std::set_intersection` they only require `T` to have a `operator<` with some wacky template meta programming.  In the [documentation](http://www.cplusplus.com/reference/algorithm/set_intersection/) it says that:
> > Two elements, a and b are considered equivalent if (!(a<b) && !(b<a)) or if (!comp(a,b) && !comp(b,a)).
> 
> Maybe documenting this requirement plus the `DRAKE_ASSERT` suggested by @jwnimmer-tri would be enough given we do not have (I think) the machinery to require a type having a given operator?
> 
</details>
I think a good fix would be doxygen of `@tparam T must be an integral type` (perhaps more verbose / better sentenced / etc.) and then a compile-time check like http://en.cppreference.com/w/cpp/types/is_integral. In that case, I don't think a `DRAKE_ASSERT` is needed.  I had thought that this PR was intended up support a wide variety of `T` values such as `int`, `double`, `TaylorVarXd`, etc. etc.  If it's really just meant for integers, than enforcing that means my top-post concern is moot.

Background: The code as of r1 left the type of `T` wide open, so it requires much thought to develop, review, and test that the routine is safe no matter what `T` the calling code provides.  This is why, all else being equal, you should write straightforward routines such as `HaveIntersection(vector<int>)` instead of pre-emptively generalizing it to support any `T` -- nevermind just keeping the routine hidden within the collision directory in the first place, or whether the performance gain from this routine merits all of this work (versus the intersection logic the STL offers).  But that's getting a bit far away from this PR's direct defects.  If the PR gets the `static_assert(is_integral...)` fix noted above, it's good enough for this time around.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/common/sorted_vectors_have_intersection.h, line 34 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9qJYEDH5qpO90teoY:-KNE36EjHY1zrkV00xg6:-1318987429) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L34)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Per comment below this documentation should at least warn about problems with floating point types, if not ban them altogether.
</details>
I documented T to be of integral type.

---

*[drake/common/sorted_vectors_have_intersection.h, line 62 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KN9oJDA6SNQaUhYFsdk:-KNDJFDrFK0ZLEknX_wt:-863489418) ([raw file](https://github.com/robotlocomotion/drake/blob/c73ea444f17cbb45aeecc53066131a53e8d2a171/drake/common/sorted_vectors_have_intersection.h#L62)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think a good fix would be doxygen of `@tparam T must be an integral type` (perhaps more verbose / better sentenced / etc.) and then a compile-time check like http://en.cppreference.com/w/cpp/types/is_integral. In that case, I don't think a `DRAKE_ASSERT` is needed.  I had thought that this PR was intended up support a wide variety of `T` values such as `int`, `double`, `TaylorVarXd`, etc. etc.  If it's really just meant for integers, than enforcing that means my top-post concern is moot.
> 
> Background: The code as of r1 left the type of `T` wide open, so it requires much thought to develop, review, and test that the routine is safe no matter what `T` the calling code provides.  This is why, all else being equal, you should write straightforward routines such as `HaveIntersection(vector<int>)` instead of pre-emptively generalizing it to support any `T` -- nevermind just keeping the routine hidden within the collision directory in the first place, or whether the performance gain from this routine merits all of this work (versus the intersection logic the STL offers).  But that's getting a bit far away from this PR's direct defects.  If the PR gets the `static_assert(is_integral...)` fix noted above, it's good enough for this time around.
</details>
There was quite a bit of discussion in #2299 about whether to implement this method or simply use set_intersection. The consensus was to implement a more "vector specific" method in order to get a very efficient implementation to be used by collision groups (cliques we decided to call them). 
At the time we decided to drop this code in common in case someone else could find it useful though I agree it could also be hidden with the collision code. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed:  one-and-a-half comments.

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/sorted_vectors_have_intersection.h, line 28 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNE7OLpURADpySmwPOV:-KNE7OLqkH0ADZZgYc15:366701934) ([raw file](https://github.com/robotlocomotion/drake/blob/32684641d995e7323cf9a7ea894c0d4c8a88cc96/drake/common/sorted_vectors_have_intersection.h#L28)):*
> ```objc
> Entries can be repeated as long as they are sorted.
> For vector `a` of size \f$ N_a \f$  and vector `b` of size \f$ N_b \f$ the
> complexity is at most \f$\mathcal{O}(N_a + N_b)\f$. The algorithm executes in
> ```

BTW, personally I think this comment was better using ASCII-math than Latex. The reason is that now it is only readable in Doxygen while before it was also readable by programmers looking in the header files. For a programmer's tool like this one I think most access will be through the header file. The big benefit of Markdown-style text is that the plain text looks fine but looks even better when formatted. Using Latex here loses that feature.

---

*[drake/common/sorted_vectors_have_intersection.h, line 43 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNE6Kjn6XwWC97fRqYk:-KNE6Kjn6XwWC97fRqYl:-639635977) ([raw file](https://github.com/robotlocomotion/drake/blob/32684641d995e7323cf9a7ea894c0d4c8a88cc96/drake/common/sorted_vectors_have_intersection.h#L43)):*
> ```objc
>   typename std::vector<T>::const_iterator bi = b.begin();
> 
>   static_assert(std::is_integral<T>::value,
> ```

Sorry to add this thought so late but it just occurred to me -- pointers should be allowed here also. They are also represented as integers and it makes sense to ask if two lists contain the same in-memory object. std::is_pointer<T>::value` is available (message should be updated too).  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KNE8OMJgC65M9t-UZUH:-1096714668)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/sorted_vectors_have_intersection.h, line 28 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNE7OLpURADpySmwPOV:-KNECRfspfMJZjzsFFOl:485991684) ([raw file](https://github.com/robotlocomotion/drake/blob/32684641d995e7323cf9a7ea894c0d4c8a88cc96/drake/common/sorted_vectors_have_intersection.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, personally I think this comment was better using ASCII-math than Latex. The reason is that now it is only readable in Doxygen while before it was also readable by programmers looking in the header files. For a programmer's tool like this one I think most access will be through the header file. The big benefit of Markdown-style text is that the plain text looks fine but looks even better when formatted. Using Latex here loses that feature.
</details>
Went back to old style ASCII. I was just playing with it. I guess it is ok for this simple documentation and I will save Latex formatting for when I need to supply a good chunk of equations. 

---

*[drake/common/sorted_vectors_have_intersection.h, line 43 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNE6Kjn6XwWC97fRqYk:-KNECbxwmKsieobJ-pF3:-842969205) ([raw file](https://github.com/robotlocomotion/drake/blob/32684641d995e7323cf9a7ea894c0d4c8a88cc96/drake/common/sorted_vectors_have_intersection.h#L43)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Sorry to add this thought so late but it just occurred to me -- pointers should be allowed here also. They are also represented as integers and it makes sense to ask if two lists contain the same in-memory object. std::is_pointer<T>::value` is available (message should be updated too).  
</details>
Updated to support pointer types and added a corresponding unit test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, just waiting for @jwnimmer-tri to ack.

---

Reviewed 1 of 4 files at r4, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/test/sorted_vectors_have_intersection_test.cc, line 78 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNEd7zuwKQTt2bfmAo0:-KNEd7zuwKQTt2bfmAo1:-323694544) ([raw file](https://github.com/robotlocomotion/drake/blob/0bfe44a0e750a7bef66961ce25a7306c606fc8f9/drake/common/test/sorted_vectors_have_intersection_test.cc#L78)):*
> ```cpp
>   vector<double*> pointer_set2 = vector<double*>{ds + 1, ds + 2, ds + 3};
>   // pointer_set1 intersects pointer_set2 (pointer to ds[3]).
>   EXPECT_TRUE(SortedVectorsHaveIntersection(pointer_set1, pointer_set2));
> ```

BTW, Nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KNEdIy8NEnMXHQy2z2b:727336004)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 4 files at r4, 1 of 2 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/common/sorted_vectors_have_intersection.h, line 39 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNEqWm2RGKnlSDimR5K:-KNEqWm3XmYEsdn1NVeE:515378461) ([raw file](https://github.com/robotlocomotion/drake/blob/0bfe44a0e750a7bef66961ce25a7306c606fc8f9/drake/common/sorted_vectors_have_intersection.h#L39)):*
> ```objc
> bool SortedVectorsHaveIntersection(const std::vector<T>& a,
>                                    const std::vector<T>& b) {
>   typename std::vector<T>::const_iterator ai = a.begin();
> ```

BTW Consider moving these two lines after both assertions.  It's a bit odd to begin the logic of a method, only then to retreat and confirm the preconditions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[drake/common/sorted_vectors_have_intersection.h, line 39 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNEqWm2RGKnlSDimR5K:-KNHk6plUcpnb6mh9Eiz:1225171909) ([raw file](https://github.com/robotlocomotion/drake/blob/0bfe44a0e750a7bef66961ce25a7306c606fc8f9/drake/common/sorted_vectors_have_intersection.h#L39)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider moving these two lines after both assertions.  It's a bit odd to begin the logic of a method, only then to retreat and confirm the preconditions.
</details>
Good observation. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 1 of 4 files at r4, 1 of 2 files at r6, 1 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Left a few comments. I assume this'll be followed up with another PR that support non-sorted vectors (assuming such a method is useful)?

---

Reviewed 1 of 5 files at r1, 1 of 4 files at r4, 1 of 2 files at r6, 1 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[drake/common/sorted_vectors_have_intersection.h, line 11 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI1L1EQeB_qv9jUK_3:-KNI1L1EQeB_qv9jUK_4:1705211807) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L11)):*
> ```objc
> namespace drake {
> 
> /** Checks for non-empty intersection of two sorted `std::vector`'s.
> ```

Consider adding "the existence of a" after "for".

---

*[drake/common/sorted_vectors_have_intersection.h, line 27 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI1aelPWdSEnDSl4L2:-KNI1aelPWdSEnDSl4L3:-1493030298) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L27)):*
> ```objc
> 
> Entries can be repeated as long as they are sorted. For vector `a` of size `Na`
> and vector `b` of size `Nb` the complexity is at most Order(Na + Nb).
> ```

BTW, isn't is canonical just to say `O(Na + Nb)`?

---

*[drake/common/sorted_vectors_have_intersection.h, line 29 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI1m15Bt9KEXvw8QLj:-KNI1m15Bt9KEXvw8QLk:-2062228639) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L29)):*
> ```objc
> and vector `b` of size `Nb` the complexity is at most Order(Na + Nb).
> The algorithm executes in constant time for vectors with disjoint entries.
> An example of worst case is given below:
> ```

Missing "the" before "worst".

Ideally add "scenario" after "case".

---

*[drake/common/sorted_vectors_have_intersection.h, line 44 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI1zxGtI2NflK2MOi-:-KNI1zxGtI2NflK2MOi0:-1426176288) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L44)):*
> ```objc
> 
>   // Checks the precondition that the lists are sorted, only in debug builds.
>   // This means O(n) performance in Debug builds but could also catch some very
> ```

Missing "it" before "could".

---

*[drake/common/sorted_vectors_have_intersection.h, line 71 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI2RnNDW1pNVCPNMiR:-KNI2RnNDW1pNVCPNMiS:2042847433) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L71)):*
> ```objc
>     }
>   }
>   // One of the ranges run out of elements before a match was found.
> ```

"run" should be "ran"

---

*[drake/common/test/sorted_vectors_have_intersection_test.cc, line 12 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI2caJAbBrnrT_y2_o:-KNI2caJAbBrnrT_y2_p:-387078177) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/test/sorted_vectors_have_intersection_test.cc#L12)):*
> ```cpp
> namespace {
> 
> // Test with std::vector specific SortedVectorsHaveIntersection.
> ```

What does "specific SortedVectorHaveIntersection" mean in the above sentence?

---

*[drake/common/test/sorted_vectors_have_intersection_test.cc, line 17 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI2mjxlBCTf2Qew-Iw:-KNI2mjxlBCTf2Qew-Ix:1568442965) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/test/sorted_vectors_have_intersection_test.cc#L17)):*
> ```cpp
>   // sorted.
>   // An std::vector can have non sorted repeated elements. Therefore the user
>   // must ensure that the previous conditions are met.
> ```

What are the above-mentioned "previous conditions"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KNI30d8Xjy5YhxevWus:1083674813)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/common/sorted_vectors_have_intersection.h, line 11 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI1L1EQeB_qv9jUK_3:-KNI58EEw0TMpD5AJQLN:-2061117095) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Consider adding "the existence of a" after "for".
</details>
Considered.

---

*[drake/common/sorted_vectors_have_intersection.h, line 29 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI1m15Bt9KEXvw8QLj:-KNI5lDruIO67QKUT2BM:-1439629900) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L29)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing "the" before "worst".
> 
> Ideally add "scenario" after "case".
</details>
I like it.

---

*[drake/common/sorted_vectors_have_intersection.h, line 44 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI1zxGtI2NflK2MOi-:-KNI5zA5I0kVhLRk3x7V:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L44)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing "it" before "could".
</details>
Done.

---

*[drake/common/sorted_vectors_have_intersection.h, line 71 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI2RnNDW1pNVCPNMiR:-KNI61r6YFulLr61w-bD:1012476393) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/sorted_vectors_have_intersection.h#L71)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "run" should be "ran"
</details>
Nice catch. Thanks!

---

*[drake/common/test/sorted_vectors_have_intersection_test.cc, line 12 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI2caJAbBrnrT_y2_o:-KNI6FWJtgi0HSkkPbWK:2066738674) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/test/sorted_vectors_have_intersection_test.cc#L12)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> What does "specific SortedVectorHaveIntersection" mean in the above sentence?
</details>
In a previous version this algorithm would work on any kind of container. Then we realized that a vector specific version would be  more efficient. 
That comment refers to the one test performed with vectors (there was one with sets). So now it doesn't mean anything. Removed.

---

*[drake/common/test/sorted_vectors_have_intersection_test.cc, line 17 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-KNI2mjxlBCTf2Qew-Iw:-KNI72hSjQdIFdiX_Zho:537919725) ([raw file](https://github.com/robotlocomotion/drake/blob/1ec3f87e72c39cf46ce6c04ec2f208819ce2550e/drake/common/test/sorted_vectors_have_intersection_test.cc#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> What are the above-mentioned "previous conditions"?
</details>
Again comment from previous version. Now there is only one condition which is that entries must be sorted. Updated comment. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 4 files at r4, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2863#-:-KNJFWFKrYrFCq6kDDpx:1417088721)*
<!-- Sent from Reviewable.io -->
